aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -567,5 +567,260 @@ Tensor& index_add_cuda_(Tensor & self, int64_t dim, const Tensor & index, const
+    int selfSelectDim = selfInfo.collapseDims(dim);
+    selfInfo.reduceDim(selfSelectDim);
+
+    cuda::detail::TensorInfo<int64_t, unsigned int> indicesInfo = cuda::detail::getTensorInfo<int64_t, unsigned int>(index, /*positiveDim = */ true);
+    indicesInfo.collapseDims();
+
+    // A reasonable choice for when to have each thread iterate over
+    // indices to choose
+    if (numIndices <= 16) {
+      if (outInfo.dims == 1 && selfInfo.dims == 1 && indContig) {
+        SMALL_INDEX(scalar_t, unsigned int, 1, 1, -2);
+      } else if (outInfo.dims == 2 && selfInfo.dims == 2 && indContig) {
+        SMALL_INDEX(scalar_t, unsigned int, 2, 2, -2);
+      } else if (outInfo.dims == 3 && selfInfo.dims == 3 && indContig) {
+        SMALL_INDEX(scalar_t, unsigned int, 3, 3, -2);
+      } else {
+        SMALL_INDEX(scalar_t, unsigned int, -1, -1, -1);
+      }
+    } else {
+      bool indexIsMajor = indexShouldBeMajor(outInfo, outSelectDim);
+
+      if (outInfo.dims == 1 && selfInfo.dims == 1 && indContig) {
+        LARGE_INDEX(scalar_t, unsigned int, 1, 1, -2, true);
+      } else if (outInfo.dims == 2 && selfInfo.dims == 2 && indContig) {
+        if (indexIsMajor) {
+          LARGE_INDEX(scalar_t, unsigned int, 2, 2, -2, true);
+        } else {
+          LARGE_INDEX(scalar_t, unsigned int, 2, 2, -2, false);
+        }
+      } else if (outInfo.dims == 3 && selfInfo.dims == 3 && indContig) {
+        if (indexIsMajor) {
+          LARGE_INDEX(scalar_t, unsigned int, 3, 3, -2, true);
+        } else {
+          LARGE_INDEX(scalar_t, unsigned int, 3, 3, -2, false);
+        }
+      } else {
+        LARGE_INDEX(scalar_t, unsigned int, -1, -1, -1, true);
+      }
+    }
+  } else {
+    cuda::detail::TensorInfo<scalar_t, uint64_t> outInfo =
+      cuda::detail::getTensorInfo<scalar_t, uint64_t>(out, /*positiveDim = */ true);
+    int outSelectDim = outInfo.collapseDims(dim);
+    outInfo.reduceDim(outSelectDim);
+
+    cuda::detail::TensorInfo<scalar_t, uint64_t> selfInfo = cuda::detail::getTensorInfo<scalar_t, uint64_t>(self, /*positiveDim = */ true);
+    int selfSelectDim = selfInfo.collapseDims(dim);
+    selfInfo.reduceDim(selfSelectDim);
+
+    cuda::detail::TensorInfo<int64_t, uint64_t> indicesInfo = cuda::detail::getTensorInfo<int64_t, uint64_t>(index, /*positiveDim = */ true);
+    indicesInfo.collapseDims();
+
+    LARGE_INDEX(scalar_t, uint64_t, -1, -1, -1, true);
+  }
+#undef SMALL_INDEX
+#undef LARGE_INDEX
+}
+} // anonymous namespace
+
+Tensor& index_select_out_cuda(Tensor& out, const Tensor& self, long dim,
+                              const Tensor& index) {
+  static constexpr int MAX_INDEX_SELECT_DIMS = 25;
+  static constexpr string_view DIM_WARNING =
+    "Tensor too large or too many (> 25) dimensions";
+
+  TORCH_CHECK(out.device() == self.device(), "Input and output must be on the "
```

Comment:
Do all the tensors need to be on the current GPU?
```suggestion
  TORCH_CHECK(at::cuda::check_device({out, self, input}), "Input and output must be on the "
```

Reply:
I'm not sure, I think they just need to be on the same device. There is no variable `input`.

Reply:
>I'm not sure, I think they just need to be on the same device. 

Well you're replacing `THCTensor_(checkGPU)` which does require current GPU.
https://github.com/pytorch/pytorch/blob/f9eb8824f1e1cd0d73275afaeb5c81fc4b3f0de1/aten/src/THC/generic/THCTensor.cpp#L446

>There is no variable `input`.

Typo, I meant `index`.

Reply:
Thanks, I didn't know that